dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+	} else if attr.NULL != nil {
+		// Skip, if not present, it means nullable.
+	} else if len(attr.L) != 0 {
+		incCount(s, typeList)
+	} else if len(attr.M) != 0 {
+		incCount(s, typeMap)
+	} else if len(attr.SS) != 0 {
+		incCount(s, typeStringSet)
+	} else if len(attr.NS) != 0 {
+		if int64Parsable(*attr.NS[0]) {
+			incCount(s, typeNumberIntSet)
+		} else {
+			incCount(s, typeNumberFloatSet)
+		}
+	} else if len(attr.BS) != 0 {
+		incCount(s, typeBinarySet)
+	} else {
+		log.Printf("Invalid DynamoDB data type: %v - %v", attrName, attr)
+	}
+}
+
+type statItem struct {
+	Type  string
+	Count int64
+}
+
+func (s *dynamoDBSchema) inferDataTypes(stats map[string]map[string]int64, rows int64) {
+	if s.ColumnTypes == nil {
+		s.ColumnTypes = make(map[string]colType)
+	}
+
+	for col, countMap := range stats {
+		var statItems, candidates []statItem
+		var presentRows, normRows int64
+		for k, v := range countMap {
+			presentRows += v
+			if float64(v)/float64(rows) <= errThreshold {
+				// If the percentage is less than the error threshold, then
+				// this data type has a high chance to be mistakenly inserted
+				// and we should discard it.
+				continue
+			}
+			statItems = append(statItems, statItem{Type: k, Count: v})
+			normRows += v
+		}
+		if normRows == 0 {
+			log.Printf("Skip column %v with no data records", col)
+			continue
+		}
+
+		nullable := float64(rows-presentRows)/float64(rows) > errThreshold
+
+		for _, si := range statItems {
+			if float64(si.Count)/float64(normRows) > conflictThreshold {
+				// If the normalized percentage is greater than the conflicting
+				// threshold, we should consider this data type as a candidate.
+				candidates = append(candidates, si)
+			}
+		}
+
+		s.ColumnNames = append(s.ColumnNames, col)
+		if len(candidates) == 1 {
+			s.ColumnTypes[col] = colType{Name: candidates[0].Type, Nullable: nullable}
+		} else {
+			// If there is no any candidate or more than a single candidate,
+			// this column has a significant conflict on data types and then
+			// defaults to a String type.
+			s.ColumnTypes[col] = colType{Name: typeString, Nullable: nullable}
+		}
+	}
+}
+
+func (s *dynamoDBSchema) genericSchema() schema.Table {
+	colDefs := make(map[string]schema.Column)
+
+	for _, colName := range s.ColumnNames {
+		colType := s.ColumnTypes[colName]
+		colDef := schema.Column{
+			Name:    colName,
+			Type:    schema.Type{Name: colType.Name},
+			NotNull: !colType.Nullable,
+		}
+		colDefs[colName] = colDef
+	}
+
+	// Sort column names in increasing order.
+	colNames := make([]string, len(s.ColumnNames))
+	copy(colNames, s.ColumnNames)
+	sort.Strings(colNames)
+
+	// The order of primary keys is important.
+	var schemaPKeys []schema.Key
+	for _, colName := range s.PrimaryKeys {
+		schemaPKeys = append(schemaPKeys, schema.Key{Column: colName})
+		colDef := colDefs[colName]
+		colDefs[colName] = schema.Column{
+			Name:    colName,
+			Type:    colDef.Type,
+			NotNull: true,
```

Comment:
Just checking: in dynamoDB, NULL is not a legal value for a key.

Reply:
No. You can't use a column with NULL type for a primary key. 

```bash
aws dynamodb create-table --table-name Migration \                                                                             
                              --attribute-definitions AttributeName=Username,AttributeType=NULL \
                              --key-schema AttributeName=Username,KeyType=HASH \
                              --provisioned-throughput ReadCapacityUnits=5,WriteCapacityUnits=5

An error occurred (ValidationException) when calling the CreateTable operation: 1 validation error detected: Value 'NULL' at 'attributeDefinitions.1.member.attributeType' failed to satisfy constraint: Member must satisfy enum value set: [B, N, S]
```

It only allows String, Number, and Bytes as keys and I guess because they are hashable?

Reply:
Thanks!